# HG changeset patch
# User Dalimil Hajek <dalimilhajek@gmail.com>
Bug 1281709 - Reps: Implement Rep for DOM elements. r=lclark


diff --git a/devtools/client/shared/components/reps/html-element.js b/devtools/client/shared/components/reps/html-element.js
new file mode 100644
index 0000000..2c811a4
--- /dev/null
+++ b/devtools/client/shared/components/reps/html-element.js
@@ -0,0 +1,123 @@
+/* -*- indent-tabs-mode: nil; js-indent-level: 2 -*- */
+/* vim: set ft=javascript ts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use strict";
+
+// Make this available to both AMD and CJS environments
+define(function (require, exports, module) {
+  // Dependencies
+  const React = require("devtools/client/shared/vendor/react");
+  const { createFactories, isGrip, cropMultipleLines } = require("./rep-utils");
+  const { ObjectLink } = createFactories(require("./object-link"));
+  const { StringRep } = require("./string");
+
+  // Shortcuts
+  const DOM = React.DOM;
+  const { rep: StringRepFactory } = createFactories(StringRep);
+
+  /**
+   * Renders an HTML element
+   */
+  const HtmlElement = React.createClass({
+    displayName: "HtmlElementRep",
+
+    propTypes: {
+      object: React.PropTypes.object.isRequired,
+      mode: React.PropTypes.string,
+    },
+
+    getTitle: function (grip) {
+      if (this.props.objectLink) {
+        return ObjectBox({},
+          this.props.objectLink({
+            object: grip
+          }, grip.class)
+        );
+      }
+      return "";
+    },
+
+    getAttributes: function(attributes) {
+      if (attributes === undefined) {
+        return "";
+      }
+      return Object.keys(attributes).map((key) => (
+        DOM.span({},
+          " ",
+          DOM.span({className: "attrTitle"},
+            key
+          ),
+          DOM.span({className: "attrEqual"},
+            "="
+          ),
+          StringRepFactory({object: attributes[key].trim()})
+        )
+      ));
+    },
+
+    getContent: function (grip) {
+      let name = grip.preview.nodeName;
+      let attributes = grip.preview.attributes;
+      let mode = this.props.mode || "short";
+
+      if (mode == "short" || mode == "tiny") {
+        let classes = (attributes && attributes.class) ?
+          "." + attributes.class.trim().split(" ").join(".") : "";
+        let id = (attributes && attributes.id) ? ("#" + attributes.id) : "";
+        return (
+          DOM.span({},
+            DOM.span({className: "nodeName"}, name),
+            cropMultipleLines(id + classes)
+          )
+        );
+      }
+
+      let objectLink = this.props.objectLink || DOM.span;
+      return (
+        DOM.span({},
+          objectLink({object: grip},
+            "<"
+          ),
+          DOM.span({className: "nodeName"},
+            name
+          ),
+          this.getAttributes(attributes),
+          objectLink({object: grip},
+            ">"
+          )
+        )
+      );
+    },
+
+    render: function () {
+      let grip = this.props.object;
+
+      return (
+        ObjectLink({className: "HtmlElement"},
+          this.getTitle(grip),
+          DOM.span({className: "objectPropValue"},
+            this.getContent(grip)
+          )
+        )
+      );
+    },
+  });
+
+  function supportsObject(object, type) {
+    if (!isGrip(object)) {
+      return false;
+    }
+
+    return (object.preview && object.preview.kind === "DOMNode");
+  }
+
+  // Exports from this module
+
+  exports.HtmlElement = {
+    rep: HtmlElement,
+    supportsObject: supportsObject
+  };
+});
diff --git a/devtools/client/shared/components/reps/moz.build b/devtools/client/shared/components/reps/moz.build
index 17e7d30..e5f3ac3 100644
--- a/devtools/client/shared/components/reps/moz.build
+++ b/devtools/client/shared/components/reps/moz.build
@@ -5,16 +5,17 @@
 # file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
 DevToolsModules(
     'array.js',
     'attribute.js',
     'caption.js',
     'date-time.js',
     'document.js',
+    'html-element.js',
     'event.js',
     'function.js',
     'grip-array.js',
     'grip.js',
     'named-node-map.js',
     'null.js',
     'number.js',
     'object-box.js',
diff --git a/devtools/client/shared/components/reps/rep.js b/devtools/client/shared/components/reps/rep.js
index e7feaf2..969e0f7 100644
--- a/devtools/client/shared/components/reps/rep.js
+++ b/devtools/client/shared/components/reps/rep.js
@@ -20,16 +20,17 @@ define(function (require, exports, module) {
   const { Number } = require("./number");
   const { ArrayRep } = require("./array");
   const { Obj } = require("./object");
 
   // DOM types (grips)
   const { Attribute } = require("./attribute");
   const { DateTime } = require("./date-time");
   const { Document } = require("./document");
+  const { HtmlElement } = require("./html-element");
   const { Event } = require("./event");
   const { Func } = require("./function");
   const { NamedNodeMap } = require("./named-node-map");
   const { RegExp } = require("./regexp");
   const { StyleSheet } = require("./stylesheet");
   const { TextNode } = require("./text-node");
   const { Window } = require("./window");
   const { ObjectWithText } = require("./object-with-text");
@@ -47,16 +48,17 @@ define(function (require, exports, module) {
     DateTime,
     TextNode,
     NamedNodeMap,
     Attribute,
     Func,
     ArrayRep,
     Document,
     Window,
+    HtmlElement,
     ObjectWithText,
     ObjectWithURL,
     GripArray,
     Grip,
     Undefined,
     Null,
     StringRep,
     Number,
diff --git a/devtools/client/shared/components/reps/reps.css b/devtools/client/shared/components/reps/reps.css
index 2cfc80c..fdb0bed 100644
--- a/devtools/client/shared/components/reps/reps.css
+++ b/devtools/client/shared/components/reps/reps.css
@@ -108,17 +108,20 @@
 /******************************************************************************/
 
 .objectLink-object .nodeName,
 .objectLink-NamedNodeMap .nodeName,
 .objectLink-NamedNodeMap .objectEqual,
 .objectLink-NamedNodeMap .arrayLeftBracket,
 .objectLink-NamedNodeMap .arrayRightBracket,
 .objectLink-Attr .attrEqual,
-.objectLink-Attr .attrTitle {
+.objectLink-Attr .attrTitle,
+.objectLink-HtmlElement .nodeName,
+.objectLink-HtmlElement .attrTitle,
+.objectLink-HtmlElement .attrEqual {
   color: var(--node-color);
 }
 
 .objectLink-object .nodeName {
   font-weight: normal;
 }
 
 /******************************************************************************/
